examples_tests/17.SimpleBulletIntegration/Camera.hpp

```diff
@@ -0,0 +1,377 @@
+					keysDown[Keys::EKA_MOVE_FORWARD] = true;
+				} else if(ev.action == ui::SKeyboardEvent::ECA_RELEASED) {
+					perActionDt[Keys::EKA_MOVE_FORWARD] -= timeDiff; 
+					keysDown[Keys::EKA_MOVE_FORWARD] = false;
+				}
+			}
+
+			if(ev.keyCode == ui::EKC_DOWN_ARROW || ev.keyCode == ui::EKC_S) {
+				if(ev.action == ui::SKeyboardEvent::ECA_PRESSED && keysDown[Keys::EKA_MOVE_BACKWARD] == false) {
+					perActionDt[Keys::EKA_MOVE_BACKWARD] += timeDiff; 
+					keysDown[Keys::EKA_MOVE_BACKWARD] = true;
+				} else if(ev.action == ui::SKeyboardEvent::ECA_RELEASED) {
+					perActionDt[Keys::EKA_MOVE_BACKWARD] -= timeDiff; 
+					keysDown[Keys::EKA_MOVE_BACKWARD] = false;
+				}
+			}
+
+			if(ev.keyCode == ui::EKC_LEFT_ARROW || ev.keyCode == ui::EKC_A) {
+				if(ev.action == ui::SKeyboardEvent::ECA_PRESSED && keysDown[Keys::EKA_MOVE_LEFT] == false) {
+					perActionDt[Keys::EKA_MOVE_LEFT] += timeDiff; 
+					keysDown[Keys::EKA_MOVE_LEFT] = true;
+				} else if(ev.action == ui::SKeyboardEvent::ECA_RELEASED) {
+					perActionDt[Keys::EKA_MOVE_LEFT] -= timeDiff; 
+					keysDown[Keys::EKA_MOVE_LEFT] = false;
+				}
+			}
+
+			if(ev.keyCode == ui::EKC_RIGHT_ARROW || ev.keyCode == ui::EKC_D) {
+				if(ev.action == ui::SKeyboardEvent::ECA_PRESSED && keysDown[Keys::EKA_MOVE_RIGHT] == false) {
+					perActionDt[Keys::EKA_MOVE_RIGHT] += timeDiff; 
+					keysDown[Keys::EKA_MOVE_RIGHT] = true;
+				} else if(ev.action == ui::SKeyboardEvent::ECA_RELEASED) {
+					perActionDt[Keys::EKA_MOVE_RIGHT] -= timeDiff; 
+					keysDown[Keys::EKA_MOVE_RIGHT] = false;
+				}
+			}
+		}
+		
+		float moveSpeed = 0.02f;
+		core::vectorSIMDf pos; pos.set(getPosition());
+		core::vectorSIMDf target = getTarget() - pos;
+		core::vector3df relativeRotation = target.getAsVector3df().getHorizontalAngle();
+	
+		// set target
+		target.set(0,0, core::max(1.f, core::length(pos)[0]), 1.f);
+		core::vectorSIMDf movedir = target;
+
+		core::matrix3x4SIMD mat;
+		{
+			core::matrix4x3 tmp;
+			tmp.setRotationDegrees(core::vector3df(relativeRotation.X, relativeRotation.Y, 0));
+			mat.set(tmp);
+		}
+		mat.transformVect(target);
+		target.makeSafe3D();
+
+		movedir = target;
+
+		movedir.makeSafe3D();
+		movedir = core::normalize(movedir);
+
+		pos += movedir * perActionDt[Keys::EKA_MOVE_FORWARD] * moveSpeed;
+
+		pos -= movedir * perActionDt[Keys::EKA_MOVE_BACKWARD] * moveSpeed;
+
+		// strafing
+
+		core::vectorSIMDf strafevect; strafevect.set(target);
+		if (leftHanded)
+			strafevect = core::cross(strafevect, upVector);
+		else
+			strafevect = core::cross(upVector, strafevect);
+
+		strafevect = core::normalize(strafevect);
+
+		pos += strafevect * perActionDt[Keys::EKA_MOVE_LEFT] * moveSpeed;
+
+		pos -= strafevect * perActionDt[Keys::EKA_MOVE_RIGHT] * moveSpeed;
+
+		// write translation
+		setPosition(pos.getAsVector3df());
+
+		// write right target
+		target += pos;
+		setTarget(target.getAsVector3df());
+
+		lastVirtualUpTimeStamp = nextPresentationTimeStamp;
```

Comment:
this needs to be outside the event processing function (imagine cameras with movement controlled by other things)